File: ruby-sinatra-js/paynet.rb
# frozen_string_literal: true

# The three gateway calls the Hosted Fields flow needs.
# https://doc.payneteasy.com/integration/api_use_cases/hosted_fields.html

require 'json'
require 'logger'
require 'net/http'
require 'time'
require 'uri'

require_relative 'oauth'
require_relative 'settings'

module Paynet
  # The Go example's http.Client has the same 30 seconds, and Next passes the same AbortSignal:
  # without one a wedged gateway holds the request, and the payer's page, indefinitely. Net::HTTP
  # needs both halves named, or it waits forever on the half left out.
  OPEN_TIMEOUT = 30
  READ_TIMEOUT = 30

  LOG = Logger.new($stderr)
  LOG.formatter = proc { |_severity, time, _program, message| "#{time.iso8601} #{message}\n" }

  # Step 1. A single-use ticket the browser exchanges for a hosted fields token. Valid for 15
  # minutes and safe to put on the page.
  def self.ephemeral_ticket
    reply = post_json("#{Settings::API_URL}/api/v4/tokenize/create-ephemeral-ticket/#{Settings::ENDPOINT_ID}")

    ticket = reply['ephemeralTicket'].to_s.strip
    return ticket unless ticket.empty?

    # A rejected request comes back as 4xx with a JSON body carrying the reason. Only that one
    # field is quoted: this reason travels on into config.js, where the browser can read it, and
    # the rest of the reply is the gateway's business and not the payer's.
    message = reply['error-message'].to_s
    raise(message.empty? ? 'no ephemeralTicket in the response' : message)
  end

  # Step 3. Charges the card behind the hosted fields token. The token replaces
  # credit_card_number, expire_month, expire_year and cvv2 — sending those is an error.
  #
  # `payer` holds what our own inputs next to the card iframes collected, as
  # firstName / lastName / email / cardPrintedName.
  def self.sale(hosted_fields_token:, client_order_id:, ip_address:, browser:, payer:)
    # 3DS 2.0 browser data, required by /api/v4/sale. It comes from the page, so it goes first and
    # every server-owned field below overwrites it — app.rb has already filtered it to the
    # documented keys, and this ordering is what keeps that a belt and not a single thread.
    params = browser.merge(
      'client_orderid' => client_order_id,
      'order_desc' => 'Hosted Fields example order',
      'amount' => Settings::ORDER_AMOUNT,
      'currency' => Settings::ORDER_CURRENCY,
      'hosted_fields_token' => hosted_fields_token,
      # Composed by the page from first and last name: the hosted fields token does not carry the
      # holder, and with the token the platform leaves it empty unless it is sent here — which
      # some acquirers do not survive.
      'card_printed_name' => payer['cardPrintedName'],
      'first_name' => payer['firstName'],
      'last_name' => payer['lastName'],
      'address1' => '100 Main st',
      'city' => 'Seattle',
      'zip_code' => '98102',
      'country' => 'US',
      'state' => 'WA',
      'phone' => '+12063582043',
      'email' => payer['email'],
      'ipaddress' => ip_address,
      'redirect_url' => Settings::REDIRECT_URL
    )

    post_json("#{Settings::API_URL}/api/v4/sale/#{Settings::ENDPOINT_ID}", params)
  end

  # Step 4. Polled until the order reaches a final status.
  def self.status(order_id:, client_order_id:)
    post_json(
      "#{Settings::API_URL}/api/v4/status/#{Settings::ENDPOINT_ID}",
      'login' => Settings::MERCHANT_LOGIN,
      'client_orderid' => client_order_id,
      'orderid' => order_id
    )
  end

  # Sends a signed command and decodes the reply.
  #
  # A rejected request — a validation error or a decline — comes back as 4xx with a JSON body, so
  # the body is decoded whatever the status: it carries the error-message for the page. Unlike
  # Python's urlopen, Net::HTTP does not raise on a 4xx, so there is nothing to catch here. Only a
  # reply that is not a JSON object at all counts as a failure of the call itself.
  def self.post_json(url, params = {})
    body, status = post(url, params)

    decoded = begin
      JSON.parse(body)
    rescue JSON::ParserError
      nil
    end

    # The body is not quoted: it reaches the page as {error}. The log line has the detail.
    raise "gateway request failed with #{status}" unless decoded.is_a?(Hash)

    decoded
  end

  # The transport. Always POST, always form-encoded, always signed.
  def self.post(url, params = {})
    uri = URI.parse(url)

    request = Net::HTTP::Post.new(uri)
    request['Content-Type'] = 'application/x-www-form-urlencoded'
    # Ask for JSON instead of the default x-www-form-urlencoded reply, which arrives as key=value
    # pairs separated by newlines. https://doc.payneteasy.com/integration/openapi.html
    request['Accept'] = 'application/vnd.pay+json'
    request['Authorization'] = OAuth.auth_header('POST', url, params)
    request.body = URI.encode_www_form(params)

    reply = Net::HTTP.start(
      uri.hostname, uri.port,
      use_ssl: uri.scheme == 'https', open_timeout: OPEN_TIMEOUT, read_timeout: READ_TIMEOUT
    ) { |http| http.request(request) }

    LOG.info("[paynet] POST #{url} -> #{reply.code}#{log_reason(reply.body)}")
    [reply.body.to_s, reply.code.to_i]
  end

  # What goes in the log beside the status code.
  #
  # Not the body: a status reply carries the card's last four digits and the holder's name, and
  # the ticket reply carries the ticket. The gateway puts everything a log needs to be useful into
  # these two fields anyway.
  def self.log_reason(body)
    decoded = begin
      JSON.parse(body.to_s)
    rescue JSON::ParserError
      nil
    end
    return ' (reply is not JSON)' unless decoded.is_a?(Hash)

    reason = ''
    order_id = log_field(decoded, 'paynet-order-id')
    reason += " order #{order_id}" unless order_id.empty?
    message = log_field(decoded, 'error-message')
    reason += " #{message}" unless message.empty?

    reason
  end

  # Renders one value of a decoded reply as a single line, whatever JSON type it arrived as.
  #
  # The gateway answers paynet-order-id as a number in a sale reply and as a string everywhere
  # else; JSON.parse keeps an integer exact, so a long one does not print as 1.2345678901e+10.
  def self.log_field(decoded, name)
    value = decoded[name]
    return '' if value.nil? || value.is_a?(Hash) || value.is_a?(Array)

    value.to_s.split.join(' ')
  end
end

File: nodejs-express-ts-js/src/paynet.ts
// Payment Gateway calls used by the Hosted Fields flow.
// https://doc.payneteasy.com/integration/api_use_cases/hosted_fields.html

import { API_URL, ENDPOINT_ID, MERCHANT_LOGIN, ORDER_AMOUNT, ORDER_CURRENCY, REDIRECT_URL } from './config.ts';
import { isJsonObject, type JsonObject, text } from './json.ts';
import { buildAuthHeader } from './oauth.ts';

// A rejected request — a validation error or a decline — comes back as 4xx with a JSON body,
// so the body is decoded whatever the status: it carries the error-message the page shows the
// payer. Only a reply that is not JSON at all counts as a failure of the call itself.
async function postJson(url: string, params?: Record<string, string>): Promise<JsonObject> {
  const { text: body, status } = await post(url, params);
  // JSON.parse is `any`. Annotating the result `unknown` is what forces the check below to
  // exist: without it every field of a reply the gateway wrote would be read as though this
  // file had promised its shape.
  let decoded: unknown;
  try {
    decoded = JSON.parse(body);
  } catch {
    // The body is not quoted: it reaches the page as {error}. The log below has the detail.
    throw new Error(`gateway request failed with ${status}`);
  }
  if (!isJsonObject(decoded)) throw new Error(`gateway request failed with ${status}`);
  return decoded;
}

async function post(url: string, params?: Record<string, string>): Promise<{ text: string; status: number }> {
  const body = params ? new URLSearchParams(params).toString() : '';
  const response = await fetch(url, {
    method: 'POST',
    // The Go example's http.Client has the same 30s, and Next passes the same AbortSignal:
    // without one a wedged gateway holds the request, and the payer's page, indefinitely.
    signal: AbortSignal.timeout(30_000),
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
      // Ask for JSON instead of the default x-www-form-urlencoded reply, which arrives as
      // key=value pairs separated by newlines. https://doc.payneteasy.com/integration/openapi.html
      Accept: 'application/vnd.pay+json',
      Authorization: buildAuthHeader('POST', url, params),
    },
    body,
  });
  const replyText = await response.text();
  console.log('[paynet] POST %s -> %d%s', url, response.status, logReason(replyText));
  return { text: replyText, status: response.status };
}

// What goes in the log beside the status code. Not the body: a status reply carries the card's
// last four digits and the holder's name, and the ticket reply carries the ticket. The gateway
// puts everything a log needs to be useful into these two fields anyway.
function logReason(body: string): string {
  let decoded: unknown;
  try {
    decoded = JSON.parse(body);
  } catch {
    return ' (reply is not JSON)';
  }
  if (!isJsonObject(decoded)) return ' (reply is not JSON)';
  const orderId = text(decoded, 'paynet-order-id') ? ` order ${text(decoded, 'paynet-order-id')}` : '';
  const reason = text(decoded, 'error-message') ? ` ${oneLine(text(decoded, 'error-message'))}` : '';
  return orderId + reason;
}

function oneLine(value: string): string {
  return value.split(/\s+/).filter(Boolean).join(' ');
}

// Step 1. Single-use ticket, valid 15 minutes, safe to embed into the page.
export async function getEphemeralTicket(): Promise<string> {
  const data = await postJson(`${API_URL}/api/v4/tokenize/create-ephemeral-ticket/${ENDPOINT_ID}`);
  const ticket = text(data, 'ephemeralTicket').trim();
  if (!ticket) throw new Error(text(data, 'error-message') || 'no ephemeralTicket in the response');
  return ticket;
}

export interface SaleRequest {
  hostedFieldsToken: string;
  clientOrderId: string;
  /** Straight off the request body, so it is a bag of unknowns rather than a Customer. */
  customer: JsonObject;
  ipaddress: string;
  browser: Record<string, string>;
}

// Step 3. A regular Sale where hosted_fields_token replaces the card parameters.
// credit_card_number, expire_month, expire_year and cvv2 must not be sent.
export async function createSale({ hostedFieldsToken, clientOrderId, customer, ipaddress, browser }: SaleRequest): Promise<JsonObject> {
  return postJson(`${API_URL}/api/v4/sale/${ENDPOINT_ID}`, {
    // 3DS 2.0 browser data, required by /api/v4/sale. It comes from the page, so it goes first
    // and every server-owned field below overwrites it — the handler already filtered it to the
    // documented keys, and this ordering is what keeps that a belt and not a single thread.
    ...browser,
    client_orderid: clientOrderId,
    order_desc: 'Hosted Fields example order',
    amount: ORDER_AMOUNT,
    currency: ORDER_CURRENCY,
    hosted_fields_token: hostedFieldsToken,
    // Composed by the page from first and last name: the hosted fields token does not carry
    // the holder, and the platform leaves it empty unless it is sent here.
    card_printed_name: text(customer, 'cardPrintedName'),
    first_name: text(customer, 'firstName'),
    last_name: text(customer, 'lastName'),
    address1: '100 Main st',
    city: 'Seattle',
    zip_code: '98102',
    country: 'US',
    state: 'WA',
    phone: '+12063582043',
    email: text(customer, 'email'),
    ipaddress,
    redirect_url: REDIRECT_URL,
  });
}

// Order status request, polled until a final status
export async function getStatus({ orderId, clientOrderId }: { orderId: string; clientOrderId: string }): Promise<JsonObject> {
  return postJson(`${API_URL}/api/v4/status/${ENDPOINT_ID}`, {
    login: MERCHANT_LOGIN,
    client_orderid: clientOrderId,
    orderid: orderId,
  });
}
